pkg/minikube/tunnel/loadbalancer_patcher.go

```diff
@@ -0,0 +1,158 @@
+	"k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/minikube/pkg/minikube/tunnel/types"
+)
+
+type LoadBalancerPatcher interface {
+	PatchServices() ([]string, error)
+	Cleanup() ([]string, error)
+}
+
+type RequestSender interface {
+	Do(request *rest.Request) (result []byte, err error)
+}
+
+type PatchConverter interface {
+	Convert(restClient rest.Interface, patch *types.Patch) *rest.Request
+}
+
+type DefaultPatcher struct {
+	coreV1Client   v1.CoreV1Interface
+	requestSender  RequestSender
+	patchConverter PatchConverter
+}
+
+func (l *DefaultPatcher) PatchServices() (managedServices []string, err error) {
+	managedServices, err = l.applyOnLBServices(func(restClient rest.Interface, svc core_v1.Service) ([]byte, error) {
+		return l.updateService(restClient, svc)
+	})
+	return
+}
+
+func (l *DefaultPatcher) Cleanup() (managedServices []string, err error) {
+	managedServices, err = l.applyOnLBServices(func(restClient rest.Interface, svc core_v1.Service) ([]byte, error) {
+		return l.cleanupService(restClient, svc)
+	})
+	return
+}
+
+func (l *DefaultPatcher) applyOnLBServices(action func(restClient rest.Interface, svc core_v1.Service) ([]byte, error)) ([]string, error) {
+	services := l.coreV1Client.Services("")
+	serviceList, e := services.List(metav1.ListOptions{})
+	if e != nil {
+		return nil, e
+	}
+	restClient := l.coreV1Client.RESTClient()
+
+	var managedServices []string
+
+	for _, svc := range serviceList.Items {
+		if svc.Spec.Type == "LoadBalancer" {
+			managedServices = append(managedServices, svc.Name)
+			result, err := action(restClient, svc)
+			if err != nil {
+				logrus.Errorf("%s", result)
+				logrus.Errorf("error patching service %s/%s: %s", svc.Namespace, svc.Name, err)
+				continue
+			}
+		}
+	}
+	return managedServices, nil
+}
+func (l *DefaultPatcher) updateService(restClient rest.Interface, svc core_v1.Service) (result []byte, err error) {
+	clusterIP := svc.Spec.ClusterIP
+	ingresses := svc.Status.LoadBalancer.Ingress
+	if len(ingresses) == 0 || (len(ingresses) == 1 && ingresses[0].IP != clusterIP) {
+		jsonPatch := fmt.Sprintf(`[{"op": "add", "path": "/status/loadBalancer/ingress", "value":  [ { "ip": "%s" } ] }]`, clusterIP)
+		patch := &types.Patch{
+			Type:         k8s_types.JSONPatchType,
+			ResourceName: svc.Name,
+			NameSpaceSet: true,
+			NameSpace:    svc.Namespace,
+			Subresource:  "status",
+			Resource:     "services",
+			BodyContent:  jsonPatch,
+		}
+		request := l.patchConverter.Convert(restClient, patch)
+		result, err = l.requestSender.Do(request)
+		logrus.Infof("Patched %s with IP %s", svc.Name, clusterIP)
+		return
+	} else {
+		return nil, nil
+	}
+}
+
+func (l *DefaultPatcher) cleanupService(restClient rest.Interface, svc core_v1.Service) (result []byte, err error) {
+	ingresses := svc.Status.LoadBalancer.Ingress
+	if len(ingresses) > 0 {
+		jsonPatch := fmt.Sprintf(`[{"op": "remove", "path": "/status/loadBalancer/ingress" }]`)
```

Comment:
Is this a constant? Why are we using sprintf with no arguments?

Reply:
copy paste error, good catch